### ISSUE TITLE
tf.nn.sparse_softmax_cross_entropy_with_logits silently returns NaN for an out of range label under jit_compile=True, while eager and autoclustering correctly raise [Original/Qwen]

### ISSUE THREAD

**Issue by MRiffiAslett** [tensorflow/tensorflow#126005]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.14.0 (built from source, snapshot from May 2023)

### Custom code

Yes

### OS platform and distribution

Ubuntu 24.04.4 LTS (x86_64)

### Python version

3.10

### Bazel version

7.7.0, invoked via Bazelisk

### GCC/compiler version

clang 17

### CUDA/cuDNN version

CPU only environment

### GPU model and memory

CPU only environment

### Current behavior?

I ran the same sparse softmax cross entropy call three ways: plain eager, `tf.function(jit_compile=True)`, and `tf.function` under `TF_XLA_FLAGS=--tf_xla_auto_jit=2` (autoclustering). One of the label values is out of range for the number of classes, and the three modes do not agree on what happens.

`labels` is `[1, 2, 0]` and `logits` only has 2 classes, so label `2` is invalid. Eager correctly rejects it:

```
eager : InvalidArgumentError Received a label value of 2 which is outside the valid range of [0, 2). Label values: 1 2 0
```

Autoclustering agrees:

```
ac : InvalidArgumentError Received a label value of 2 which is outside the valid range of [0, 2). Label values: 1 2 0
```

`jit_compile=True` does not:

```
xla : [0.7443967 nan 1.037488]
```

Since `2` is genuinely out of range for a 2 class problem, I expected `jit_compile=True` to raise the same error eager and autoclustering do, not compile the call and hand back a loss vector with a NaN sitting where the invalid label was. Anything built on top of this loss, gradients, an optimizer step, would pick up that NaN silently instead of getting a clear error to debug from.

This looks close to the kind of thing #112068 and #117658 describe, where XLA skips a bounds check that eager and graph tracing both enforce, just showing up here through a different op.

### Standalone code to reproduce the issue

```python
import os
os.environ["TF_XLA_FLAGS"] = "--tf_xla_auto_jit=2"

import tensorflow as tf

print("tf    :", tf.__version__)

labels = tf.constant([1, 2, 0])
logits = tf.constant([[0.2, 0.1], [0.6, 0.2], [0.1, 0.7]])


def loss(labels, logits):
    return tf.nn.sparse_softmax_cross_entropy_with_logits(labels, logits)


def show(name, thunk):
    try:
        print(name, ":", thunk().numpy())
    except Exception as e:
        msg = str(e).splitlines()
        short = next((l for l in msg if "label value" in l or "Received" in l), msg[0])
        print(name, ":", type(e).__name__, short.strip())


show("eager", lambda: loss(labels, logits))
show("xla  ", lambda: tf.function(loss, jit_compile=True)(labels, logits))
show("ac   ", lambda: tf.function(loss)(labels, logits))
```

### Relevant log output

```
tf    : 2.14.0
eager : InvalidArgumentError {{function_node __wrapped__SparseSoftmaxCrossEntropyWithLogits_device_/job:localhost/replica:0/task:0/device:CPU:0}} Received a label value of 2 which is outside the valid range of [0, 2).  Label values: 1 2 0 [Op:SparseSoftmaxCrossEntropyWithLogits] name:
xla   : [0.7443967       nan 1.037488 ]
ac    : InvalidArgumentError Received a label value of 2 which is outside the valid range of [0, 2).  Label values: 1 2 0
```

**Comment by MRiffiAslett** [comment]
Follow up after reading the actual lowering (tensorflow/compiler/tf2xla/kernels/softmax_op.cc, SparseSoftmaxXentWithLogitsOp::Compile).

The NaN under jit_compile=True is not a missing check, it looks deliberate. There is a comment right above the relevant code saying an out of range index gets folded into NaN "to obey the interface contract" of the op. My read is this is there because the same lowering also has to work when it ends up compiled for a device where a kernel cannot synchronously raise back to the caller the way the CPU eager kernel does, so NaN is used as a safe poison value instead.

So the actual disagreement is between the two XLA paths, not between XLA and eager as I first framed it: autoclustering raises here, plain jit_compile=True returns NaN, and only one of those matches what the file's own comment says is meant to happen. I do not think forcing jit_compile=True to raise is the right fix since that would go against what reads as an intentional design choice, so I am not opening a PR for this one.

Leaving this open since the split between the two XLA modes still seems worth a look from someone with more context on why autoclustering ends up on a different path here, but wanted to correct my own framing rather than leave the original report standing as is.
